### ISSUE TITLE
LFX Workspace: Create a Japanese translation agent for CNCF videos

### ISSUE THREAD

**Issue by ainozaki** [WasmEdge/WasmEdge#4047]
### Project Title

Create a Japanese translation agent for CNCF videos

### Motivation

WasmEdge is a cross-platform and lightweight runtime for AI models. The WasmEdge community has created a video translation agent combining LLM(translation), whisper (voice-to-text), and GPT-SoVITS (text-to-voice). The translation agent has already translated various English videos into Chinese and released them as [[VideoLangua.com](http://videolangua.com/)](http://videolangua.com/). This project aims to create a Japanese translation agent specifically tailored to CNCF content.

### Expected Outcome

- Evaluate the quality of existing video translation to Japanese.
- Create a Japanese translation agent that consists of voice-to-text conversion, translation, and text-to-voice conversion.
- Translate CNCF-related videos into Japanese.

### Details

This project consists of following components:

- Evaluate the performance of existing translation agents for Japanese and identify key challenges
- Develop voice-to-text conversion using whisper, optimized for CNCF contents
- Create a natural English-to-Japanese translation of CNCF terms using suitable LLM models and deploy it on LlamaEdge
- Develop text-to-voice conversion using GPT-SoVITS, with a special focus on accurately pronouncing CNCF technical terms in Japanese.
- Apply the translation agent to CNCF videos.

### Milestones

- Week 1
    - [ ] Discuss the project timeline.
    - [ ] Setup development environment
        - LlamaEdge
        - Whisper (voice to text)
        - Llama (translate)
        - GPT-SoVITS (text to voice)
    - [ ] Identify target videos that have not been translated into Japanese.
        - CNCF videos
        - GOSIM Paris 2025
- Week 2: Evaluate Existing Video Translator
    - [ ] Use VideoLangua to translate selected videos into Japanese and evaluate the quantity.
- Week 3: Voice-to-English text
    - [ ] Use whisper to extract a time-stamped English transcript from a sample of CNCF videos.
    - [ ] Develop whisper prompts that are suitable for CNCF technical content.
- Week4-5: English text to Japanese text
    - [ ] Evaluate and select a Japanese language LLM that might give better translation results.
    - [ ] Develop LLM prompts that are suitable for CNCF technical content.
    - [ ] Run it as a LlamaEdge service.
- Week 6-8: Japanese text to Voice
    - [ ] Train Japanese TTS actor models for GPT-SoVITS using PyTorch.
    - [ ] Create dictionaries for how to pronounce CNCF technical words in Japanese.
    - [ ] Evaluate the synthesized Japanese voice.
- Week 9-10: Apply to real videos
    - [ ] Produce the entire selected video in week1
    - [ ] Update the translation agent.
- Week 11: Create UI
    - [ ] Create a Japanese version of the VideoLangua UI
- Week 12:
    - [ ] Document the project

## Reference Links
- https://github.com/WasmEdge/WasmEdge/issues/3986

**Comment by ainozaki** [comment]
Hello @juntao  @alabulei1 @MileyFu
I've created the project timeline. I'd like to refine it at today's meeting!

**Comment by ainozaki** [comment]
## Week1 (March 4 - 11) Updates
- Run Llama on LlamaEdge
- Run whisper on LlamaEdge
- Investigate domain-specific LLM

For details, I'm updating this [doc](https://docs.google.com/document/d/19Fv7zajun6DP4I6y8oQPmxWz9K6c9BU1rwHtKBAcuCE/edit?tab=t.0).


### Toughts
- I’m wondering if we can develop a CNCF domain-specific LLM through continual pre-training on Llama, rather than relying on prompt engineering? It would help not only translation tasks but also the understanding of large CNCF softwares, which often exceeds the limits of human comprehension. I've been surveying examples of domain-specific LLMs with Continuous Pre-Training based on Llama, and would like to continue to consider them in terms of learning costs and data sets.
- Since I still cannot access to the re-editing page on VideoLangua (public version), I’ll run VideoLangua locally and work on data collection next week  

cc  @juntao @alabulei1 @MileyFu

**Comment by juntao** [comment]
Do you mean fine-tuning? We do not have access the to pre-training data, software, or compute clusters of these models. We can, however, to fine-tune the models they published with domain specific content and materials. 

However, our experience with fine-tuning is as follows

1 It is effective in changing the tone or style of the LLM output. For example, to make the LLM "speak like Donald Trump" or to make it "always respond in Japanese".

2 It does not "add knowledge" to the LLM. So, adding CNCF specific content is unlikely to make it better. Adding these to the prompt is far more effective. 

If you want to try fine-tuning, I suggest using unsloth

https://unsloth.ai

which I covered in my book. ;)

https://www.manning.com/liveproject/fine-tune-the-llama-model

**Comment by ainozaki** [comment]
Thank you for your explanation 😃 
> Do you mean fine-tuning? 

No, I understand Continual pre-training is defferent from fine-tuning, which is task-specific. Let me quote [1]: 

> Continual pre-training is a form of domain adaptation where additional pre-training tasks are performed on a pre-trained language model using text from the target downstream task before fine-tuning the model on that task.

Several Japanese LLMs and Japanese domain-specific LLMs (e.g., 1, 2) are developed through continual pre-training on open LLMs like Llama.
[1]https://arxiv.org/abs/2404.17790, https://huggingface.co/tokyotech-llm/Swallow-7b-hf
[2]https://arxiv.org/abs/2404.10555

I found the page on Continued Pretraining in unsloth:  
https://docs.unsloth.ai/basics/continued-pretraining

**Comment by juntao** [comment]
Hmm. I guess you can try unsloth if it can do this? Open questions:

* What is the data corpus do you want to use in this?
* Can we continue to pre-train instruct models or just the base model?
* If it is the base model, how do we instruct tune it to answer questions?

**Comment by ainozaki** [comment]
I'll give it a try.
- Q1: I'll look into it more since the dataset is very important
- Q2, Q3: It seems to be the base model. I'm checking the [sample code](https://colab.research.google.com/drive/1tEd1FrOXWMnCU9UIvdYhs61tkxdMuKZu?usp=sharing#scrollTo=B2tv8AEhPTu6), and it runs instruction fine-tuning after continued pre-training.

**Comment by alabulei1** [comment]
Hi @ainozaki,

Could you please provide an update on the progress from last week (March 10-14)?

I understand that fine-tuning a model requires significant time. In the meantime, I'd appreciate if you could:

- Test the existing models to evaluate their translation quality between English and Japanese (in both directions)
- Assess Whisper's performance in transcribing Japanese audio

I will provide you with documentation later. Looking forward to your update. Thanks.

**Comment by ainozaki** [comment]
I'm sorry for the delay and thank you for your suggestions.
As you mentioned, I started evaluating the performance of existing model on youtube-bot for translation to Japanese.
- Successfully run youtube-bot locally
- Translated some movies and found:
    - The translated Japanese is somewhat awkward. For example, Japanese often omits the subject, but the translated Japanese includes all the subjects, so it sounds redundant.
    - Occasionally, Chinese or Korean is mixed in.
        - Example: それが私たちがこの番組を如此愛する...
        - Example: その作業グループは 계속成長し、魔法のよう...
    - I tried increasing the number of translation examples and entering prompts in Japanese, but I couldn't tell if the translation was better. 
    (The quality of translation varies even with the same prompt.)
    I think we need a metric to evaluate the quality of translation.
- Investigated existing Japanese LLMs. In terms of model size and WMT20 score for Japanese-English translation tasks, I am focusing on [Llama-3.1-Swallow-8B-v0.2](https://huggingface.co/tokyotech-llm/Llama-3.1-Swallow-8B-v0.2). ref: [Japanese LLM Evaluation](https://swallow-llm.github.io/evaluation/index.ja.html?index=%22__ALL__%22&task=%5B%22Llama+3.1+Swallow+8B+Instruct+v0.3%22%2C%22Swallow-7b-instruct-v0.1%22%5D&scatter=%22__ALL__%22)
- In the next step, I will run the above model on llamaedge and translate the same movie.

**Comment by juntao** [comment]
Thank you @ainozaki A couple of updates:

1 The VideoLangua.com web site is now using LLMs optimized for different languages

* [Qwen-2.5-32b](https://huggingface.co/gaianet/Qwen2.5-32B-Instruct-GGUF) from Alibaba for Chinese translations
* [EXAONE-3.5-32b](https://huggingface.co/gaianet/EXAONE-3.5-32B-Instruct-GGUF) from LG for Korean translations

It does seem that the Chinese trained model is better at Chinese, and the Korean trained model is better at Korean. So, I think focusing on Japanese LLMs would be the right path. I suppose we can create a Gaia node for [Llama-3.3-Swallow-70B](https://huggingface.co/tokyotech-llm/Llama-3.3-Swallow-70B-Instruct-v0.4) next week for you to experiment with. @alabulei1  will let you know when the node is ready.

Another thing to try is the Gemma-3-27b LLM, which claims to be multi-lingual. You can try it

* API node (requires Gaia API key): `https://0x9fcf7888963793472bfcb8c14f4b6b47a7462f17.gaia.domains/v1`
* [Free chatbot](https://llamaedge.github.io/chatbot-ui/index?api_url=https://0x9fcf7888963793472bfcb8c14f4b6b47a7462f17.gaia.domains)

Furthermore, I think you can try OpenAI's models to better understand the current Japanese translation capabilities from close source LLMs.

2 I understand that the current translation prompts are written in English. Perhaps you can try writing the prompt in Japanese and see if it works?

**Comment by ainozaki** [comment]
Thank you for your advice😃 So I'll try:
- Use Japanese prompts
- Try Llama-3.3-Swallow-70B for Japanese trained model
- Try Gemma-3-27b for multi-lingual model
- Try OpenAI's models to compare the translation quality

**Comment by ainozaki** [comment]
# Weekly Update (March 24 - 30)
Tried some models for translation with short/long inputs and Japanese/Englisf prompts and repeat 100 times to see the trends.
The results are as follows:

| Model | Impressions | Japanese/Eglish prompt |
| --| -- | -- |
| Llama-3.3-Swallow-70B | tend to not follow the instructions[1] | Ja prompt is better in terms of following the instruction or not |
| Gemma-3-27B | good[2] | En prompt is better in terms of naturalness of translation |
| current VideoLangua | somewhat awkward[3] | En prompt is better in terms of following the instruction or not |

[1] For example, when given an English sentence about NumPy with an English prompt, the model returns NumPy source code instead of a translated Japanese sentence. Using a Japanese prompt produces the expected and natural translation. That said, this model still struggles to follow instructions when given long inputs, even with a Japanese prompt.
[2] This model produces natural translations that incorporate nuances unique to Japanese. This tendency is even more pronounced when using an English prompt. For example, other model translates "You're right" as "あなたは正しいです", which is a literal translation. In contrast, this model translates it as "おっしゃる通りです", which is more natural in Japanese.
[3] As reported last week, this model’s translations remain somewhat awkward, even when using a Japanese prompt.

**Comment by juntao** [comment]
Thanks. The current VideoLangua uses Qwen 2.5 for Japanese translation. I guess you are saying that it is no good. 

The way VideoLangua works is that it does NOT translate long paragraphs. The system recognizes each English sentence from the video, and asks it to translate the sentences one by one. 

So, the problem with following instructions over long paragraphs may not be an issue. I would update VideoLangua to use the Japanese Swallow model and see how it works within the VideoLangua system.
